### PR TITLE
Simple alexa adapter

### PR DESCRIPTION
**Pull request by trashcanmonster8**
Started creating a simple alexa adapter with sample. This adapter follows a similar pattern to the dot net adapter.

### CODE REVIEW

**Comment by szul** [comment]
@trashcanmonster8 Thanks for the contribution! I’m traveling at the moment, but can review when I get back on the 6th of January.

Can you try to fix the build failures, if they’re related to your contributions?

How closely does this align with @iMicknl’s WIP PR for an Alexa adapter? Can they be merged?

File: libraries/botbuilder-adapter-alexa/src/alexaAdapter.ts [hunk 1]
```diff
@@ -0,0 +1,150 @@
...
+    TurnContext,
+    Activity,
+    ResourceResponse,
+    WebRequest,
+    WebResponse,
+    ActivityTypes
+} from 'botbuilder';
+import { RequestEnvelope, IntentRequest, Session, interfaces, ResponseEnvelope, SessionEndedReason, SessionEndedRequest } from 'ask-sdk-model';
+import { getRequestType, getLocale, getUserId, createAskSdkError } from 'ask-sdk-core';
+import { retrieveBody } from '@botbuildercommunity/adapter-twitter';
```
Comment (szul) [inline libraries/botbuilder-adapter-alexa/src/alexaAdapter.ts]: Future note: We have `retrieveBody` in three packages now. Might be time to make a `@botbuildercommunity/util` or something.

Comment (iMicknl) [reply libraries/botbuilder-adapter-alexa/src/alexaAdapter.ts]: > How closely does this align with @iMicknl’s WIP PR for an Alexa adapter? Can they be merged?

I will have a look. This one looks more detailed than mine, so will have a look if I did some things different and what we should merge. I paused mine since @garypretty is releasing a new version of the .NET adapter and I wanted to base it on the new version. 

Thanks for your contribution @trashcanmonster8.

**Comment by iMicknl** [comment]
> How closely does this align with @iMicknl’s WIP PR for an Alexa adapter? Can they be merged?

I will have a look. This one looks more detailed than mine, so will have a look if I did some things different and what we should merge. I paused mine since @garypretty is releasing a new version of the .NET adapter and I wanted to base it on the new version. 

Thanks for your contribution @trashcanmonster8.

File: libraries/botbuilder-adapter-alexa/src/alexaAdapter.ts (hunk 1 above)
Comment (iMicknl) [reply libraries/botbuilder-adapter-alexa/src/alexaAdapter.ts]: I am not a fan of referring other packages, however having the code duplication is also not great.. Best would be to have a util package indeed or to have it in the official botframework adapter, which can be referenced.

**Comment by garypretty** [comment]
@trashcanmonster8 Thank you for the contribution.  I will reivew as soon as I can. 

As @iMicknl said he has an open PR for an Alexa adapter as well and I have been giving some thought into how this should be implemented given the recent changes to the .NET adapter.  I'll drop some thoughts here as soon as I can, but wherever possible I would like us to make use of the JS packages made available from the Alexa team themselves.  They have some articles explaining how to use the Alexa Skills Kit SDK (ask-sdk) in your own web service and so I think it makes sense to use this where possible.  A similar SDK doesn't exist for .NET. but that is why we have adopted the open source Alexa.NET package.

**Comment by garypretty** [comment]
@szul @trashcanmonster8 @iMicknl This is the article I have started looking at https://developer.amazon.com/en-US/docs/alexa/alexa-skills-kit-sdk-for-nodejs/host-web-service.html

The goal here is to have the adapter be as lightweight as possible - which is hopefully obvious from the PR currently open in the .NET repo.

**Comment by szul** [comment]
@garypretty They both are using the `ask-sdk`. I think @iMicknl was waiting to see if the .NET changes had any special sauce.

**Comment by garypretty** [comment]
@szul Ah, thanks for the heads up! I am knee deep in docs writing at the moment so I missed that. 
@trashcanmonster8 @iMicknl I would be great to see if you can consolidate your efforts. Probably not a bad idea to also check out the BotKit packages at https://github.com/howdyai/botkit/tree/master/packages and try to follow any patterns / structure you see there too.

Thanks again to you all for this - getting a JS version of the adapter is very high on the priority list!

**Comment by trashcanmonster8** [comment]
I got the checks to pass. There was a security audit for other packages that needed fixed. I went ahead and fixed those. That may need an additional review since I am not as familiar as to why those changes were needed.

So it looks like the two PR's are not too far off from one another. each can handle simple messages. They take the same fundamental approach. #95 has some verification of the requests that this PR is missing and some cleaner methods. #140 includes unit testing along with a simple hello world bot. These seem to be the big contrasts that I can see.

How would we want to move forward with this?

**Comment by garypretty** [comment]
Thanks for the update @trashcanmonster8 

One guiding principle for the JS Alexa adapter should be parity in terms of functionality / structure compared with the C# version, which is aligned with the approach taken by the SDK itself. I haven't had a chance to review yet so I don't know how far off this is. 

When I talk of parity I mean having the adapter and methods functionally do the same thing. E.g. The raw adapter should create activities based on the Alexa request types, but using middleware to transform Alexa requests into message / event activities if the developer wishes. Ultimately a developer should be able to pick up either and not see any differences unless they are language specific and unavoidable. 

For reference this is the current version of the Alexa adapter - > https://github.com/BotBuilderCommunity/botbuilder-community-dotnet/pull/179. There will be a few more changes coming but nothing major. 

@iMicknl @trashcanmonster8 it would be good to understand how much you think your implementations are currently aligned with the C# version?

I promise I will get around to reviewing this ASAP. Just have a bunch to get through before I head out to the States at the weekend.

**Comment by szul** [comment]
@garypretty I would be careful requiring absolute parity for a few reasons: 1) It might not exist (JavaScript and C#) are different and have different packages and ways of working; 2) Considering the contribution levels outside of myself, Mick, and you (on the C# end),  I don’t want to discourage contributions by forcing them to read C# code, especially since I doubt anyone contributing C# packages is going to want to read the JavaScript ones.

If the Bot Framework packages and the SDK packages from the other services (JavaScript SDK, .NET SDK) contain parity, which they should, there should be a natural alignment. If not, we should have separate documentation on suggested implementation—whether that’s in a different repo or in an issue.

Bottom line is that I’d rather have an Alexa adapter at 50% than none at all, and contributing to OSS is hard when it’s not your full time job.

My suggestion here would be to defer to @iMicknl as to whether he wants to combine the two packages himself  and do a first review, or if his time is tight, have @trashcanmonster8 combine the two, and Mick can do a first review of the code. Our pipeline should take care of most of the style differences (ESLint). Once combined, we can determine is anything is missing, and open issues for them to encourage contributions.

**Comment by garypretty** [comment]
@szul I totally agree with your points regarding contribution etc.  I absolutely would rather have a 50% contrib than nothing at all too.  I just wanted to point this out as a potential goal, for any areas that are 'quick fixes' to align the two packages.  I am also happy to pick up the packages once we have the initial versions and bring them closer in alignment with each other.

**Comment by garypretty** [comment]
The last thing I want to do is discourage anyone. I know how hard it can be to contrib in your spare time and I really hope it doesn't come across differently. I am by no means trying to put blockers in people's way or undervalue their input - I really appreciate it all, it's amazing.

**Comment by szul** [comment]
Last day of visiting family. I might just be cranky.

**Comment by trashcanmonster8** [comment]
@szul I agree with you on inherent differences between **js** and **dotnet**. Already `processActivities` behaves differently in **dotnet** by returning an `AlexaResponse` object. By contrast, the **js** method updates the response by reference. In short, there will have to be some differences in the interfaces given they already exist. But I'm not a **dotnet** expert. So I may be reading the code incorrectly.

@garypretty #140 is missing that middleware class. But since that exists already in #95 , I won't worry about creating this yet.

@iMicknl I wonder if a way to help create some merge between these pr's is to create a Converter class which can handle creating `RequestEnvelope` and `ResponseEnvelope` objects from `Activity` objects. Both adapter and middleware would consume this same class. And this would help if Alexa's api ever has breaking changes. Granted this may be over engineering.

**Comment by iMicknl** [comment]
@szul @trashcanmonster8 I will have a look at merging our packages, so we can combine the efforts.

File: libraries/botbuilder-adapter-alexa/src/alexaAdapter.ts [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+import { RequestEnvelope, IntentRequest, Session, interfaces, SessionEndedRequest } from 'ask-sdk-model';
+import { getRequestType, getLocale, getUserId, createAskSdkError } from 'ask-sdk-core';
+import { retrieveBody } from '@botbuildercommunity/adapter-twitter';
+
+/**
+ * @module botbuildercommunity/adapter-alexa
+ */
+
+export class AdapterAlexa extends BotAdapter {
+    private responses: Map<string, Activity[]> = new Map<string, Activity[]>();
```
Comment (iMicknl) [inline libraries/botbuilder-adapter-alexa/src/alexaAdapter.ts]: Why save the responses in a `Map`? For what do we need them later on?

File: libraries/botbuilder-adapter-alexa/src/alexaAdapter.ts [hunk 3]
```diff
@@ -0,0 +1,150 @@
...
+    // eslint-disable-next-line @typescript-eslint/no-unused-vars
+    public async updateActivity(_context: TurnContext, _activity: Partial<Activity>): Promise<void> {
+        throw new Error(`updateActivity is not implemented for ${ AdapterAlexa.name }`);
+    }
+
+    public async processActivity(req: WebRequest, res: WebResponse, logic: (context: TurnContext) => Promise<any>): Promise<void> {
+        const alexaRequest: RequestEnvelope = await retrieveBody(req);
+        const system: interfaces.system.SystemState = alexaRequest.context.System;
+        const requestType = getRequestType(alexaRequest);
+        const session: Session = alexaRequest.session ? alexaRequest.session : {
```
[the remark this answers is not in the document]
Comment (trashcanmonster8) [reply libraries/botbuilder-adapter-alexa/src/alexaAdapter.ts]: session is something in the alexa sdk. in the .net adapter, they use the session to add an `id` property to the conversation. my guess is so that the `conversation.id` is consistent between alexa and botframework.

File: libraries/botbuilder-adapter-alexa/src/alexaAdapter.ts (hunk 2 above)
Comment (trashcanmonster8) [reply libraries/botbuilder-adapter-alexa/src/alexaAdapter.ts]: the .net framework had this property. but something needs to be created to help the `processActivities` method and `sendActivities` methods to create an response for alexa. Alexa is expecting a their post to return a message back to the user. The botframework expects a simple 200 response with some uuid field and then sends the channel post in response.

File: libraries/botbuilder-adapter-alexa/src/alexaAdapter.ts [hunk 4]
```diff
@@ -0,0 +1,150 @@
...
+import { retrieveBody } from '@botbuildercommunity/adapter-twitter';
+
+/**
+ * @module botbuildercommunity/adapter-alexa
+ */
+
+export class AdapterAlexa extends BotAdapter {
+    private responses: Map<string, Activity[]> = new Map<string, Activity[]>();
+    // eslint-disable-next-line @typescript-eslint/no-unused-vars
+    public async continueConversation(_reference: Partial<ConversationReference>, _logic: (revocableContext: TurnContext) => Promise<void>): Promise<void> { }
```
[the remark this answers is not in the document]
Comment (trashcanmonster8) [reply libraries/botbuilder-adapter-alexa/src/alexaAdapter.ts]: From the BotAdapter class:
```
/**
     * Asynchronously resumes a conversation with a user, possibly after some time has gone by.
     *
     * @param reference A reference to the conversation to continue.
     * @param logic The asynchronous method to call after the adapter middleware runs.
     *
     * @remarks
     * This is often referred to as a _proactive notification_, the bot can proactively
     * send a message to a conversation or user without waiting for an incoming message.
     * For example, a bot can use this method to send notifications or coupons to a user.
     */
    abstract continueConversation(reference: Partial<ConversationReference>, logic: (revocableContext: TurnContext) => Promise<void>): Promise<void>;
   ```

File: libraries/botbuilder-adapter-alexa/src/alexaAdapter.ts (hunk 2 above)
Comment (trashcanmonster8) [reply libraries/botbuilder-adapter-alexa/src/alexaAdapter.ts]: @iMicknl i have some ideas about how to break apart the adapters and the classes which convert the alexa api to botframework activity. let me know if you think something like this would be helpful: https://github.com/trashcanmonster8/botbuilder-community-js/tree/alexa-activity-class/libraries/botbuilder-adapter-alexa/src

I found it made testing a lot easier too.

**Comment by trashcanmonster8** [comment]
@iMicknl what do you think about this idea: https://github.com/trashcanmonster8/botbuilder-community-js/tree/alexa-activity-class/libraries/botbuilder-adapter-alexa/src

I separated the adapter from the classes that handle the different API's. I think this would help make the adapter thinner since the logic to convert alexa API to botframework activity may be rather large.

It also made testing easier.

File: libraries/botbuilder-adapter-alexa/src/alexaAdapter.ts (hunk 2 above)
Comment (iMicknl) [reply libraries/botbuilder-adapter-alexa/src/alexaAdapter.ts]: [My approach](https://github.com/BotBuilderCommunity/botbuilder-community-js/pull/95/files#diff-28cf8d8593fe3cf4ada24a4b00c276e4R61) is slightly different and based on the [BotKit BotBuilder adapters](https://github.com/howdyai/botkit/blob/master/packages/botbuilder-adapter-slack/src/slack_adapter.ts#L571). They use the `context.turnState.set('httpBody', body)` to send a response back. What do you think? 

I am not sure what the best way is for JavaScript adapters. I will have a look at the [WeChat Typescript adapter](https://github.com/microsoft/BotFramework-WeChat/pull/3), which is a work in progress. Let's see how they are tackling this.

File: libraries/botbuilder-adapter-alexa/src/alexaAdapter.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (iMicknl) [reply libraries/botbuilder-adapter-alexa/src/alexaAdapter.ts]: Nevermind, I misread. I thought the type was different but it is just the variable name.. Good to keep it as close as possible to the `BotAdapter` class.

**Comment by iMicknl** [comment]
@trashcanmonster8 looks good! I will have a more in depth look tomorrow, but it would be good to move some code outside the adapter. The [WeChat adapter (wip)](https://github.com/microsoft/BotFramework-WeChat/tree/add/wechat-adapter-typescript/libraries/typescript/Microsoft.Bot.Builder.Adapters.WeChat/src) calls it [`MessageMapper`](https://github.com/microsoft/BotFramework-WeChat/blob/add/wechat-adapter-typescript/libraries/typescript/Microsoft.Bot.Builder.Adapters.WeChat/src/weChatMessageMapper.ts).

File: libraries/botbuilder-adapter-alexa/src/alexaAdapter.ts (hunk 2 above)
Comment (garypretty) [reply libraries/botbuilder-adapter-alexa/src/alexaAdapter.ts]: @iMicknl This is an interesting one. I think this might be where you see the differences between ported C# -> JS/TS (which I think is the case for the WeChat adapter) and those written in JS originally (i.e. BotKit, which we have just completed porting to C#).  My general steer here, at least for now, would be to follow the BotKit adapter patterns (they are certainly the ones that have had the most attention to date).  However, if you think the WeChat one is doing something better in places, then great.

Comment (iMicknl) [reply libraries/botbuilder-adapter-alexa/src/alexaAdapter.ts]: @garypretty agree. However for Twilio WhatsApp I had a look at the BotKit adapters and some features like attachment handling are not really aligned with the Bot Framework SDK. + the need for `restify.plugins.bodyParser()`, where this can easily be handled by some extra code.

I will have a look at the best practices from the BotKit adapters + the WeChat sample.

Comment (garypretty) [reply libraries/botbuilder-adapter-alexa/src/alexaAdapter.ts]: @iMicknl Absolutely great with me.  Let's follow them if / where is makes sense, but if you think there is a better or more efficient way of doing things then great. 👍 

@iMicknl @trashcanmonster8 thanks to you both for your continued effort on this. It is awesome to see and a JS / TS version of the Alexa adapter is a common ask that we hear. I'll also apologies for my lack of familiarity with JS (it's on my list to ramp up on) 👍

Comment (szul) [reply libraries/botbuilder-adapter-alexa/src/alexaAdapter.ts]: @trashcanmonster8 @iMicknl For the Twitter adapter, I followed @iMicknl's pattern of `context.turnState.set('httpBody', body)`. Botkit is a slightly different animal but Ben has some good patterns and practices we can utilize.

**Comment by iMicknl** [comment]
@trashcanmonster8 sorry for the huge delay... I have incorporated your work into #95. Could you please review and let me know if you have any comments?

**Comment by iMicknl** [comment]
I will close this one for now @trashcanmonster8, since it has been merged with #95. Thanks for your contribution!! I have added you to the credits of the other package and you have been credited in the original [Bot Framework Blog](https://blog.botframework.com/2020/02/27/bot-framework-community-focus-february-2020/).
